src/alloc/globalalloc.c

```diff
@@ -5,33 +5,154 @@ static struct allocator global_allocator;
+  {
+    metadata[i].pool_element = &pool[i];
+    metadata[i].used = false;
+    metadata[i].head = false;
+    metadata[i].allocated = -1;
+    metadata[i].index = i;
+  }
 }
 
-/** Allocates a chunk of memory of the given size in global memory */
-static void *global_malloc(size_t size) {
-    return NULL;
+/** Finds corresponding metadata node to a given pool element */
+static struct Node *find_element(void *p)
+{
+  struct Node *element = NULL;
+  for (int i = 0; i < POOL_SIZE; i++)
+  {
+    if (metadata[i].pool_element == p)
+    {
+      element = metadata + i;
+      return element;
+    }
+  }
+
+  return element;
 }
 
-/** Resizes a chunk of memory in global memory previously allocated by malloc */
-static void *global_realloc(void *p, size_t size) {
-    return NULL;
+/** Flips a chunk of data's "used" state to the opposite */
+static void switch_state(struct Node *element, size_t size)
+{
+  // Set data to opposite state
+  for (int i = 0; i < size; i++)
+  {
+    if ((element + i)->used)
+      (element + i)->used = false;
+    else
+      (element + i)->used = true;
+  }
+}
+
+/** Finds a chunk in the pool that satisfies the size requirement. Returns the head of the chunk */
+static struct Node *find_space(size_t size)
+{
+  int counter = 0;
+  while (counter < POOL_SIZE)
+  {
+    // Check if current element is head. If so, skip counter by allocated chunk size
+    if ((metadata + counter)->head)
+      counter += (metadata + counter)->allocated;
+
+    // Check if chunk size is enough space
+    bool fits = true;
+    int jump_size = 0; // Amount of bytes to "jump" ahead
+    for (int i = 0; i < size; i++)
+    {
+
+      if ((metadata + counter + i)->used)
+      {
+        fits = false;
+        jump_size = i;
+        break;
+      }
+    }
+
+    if (fits)
+      return (metadata + counter);
+    counter += jump_size;
+  }
+  return NULL;
 }
 
 /** Frees a chunk of memory in global memory previous allocated by malloc */
-static int global_free(void *p) {
-    return -1;
+static int global_free(void *p)
+{
+  struct Node *element = find_element(p);
+  if(element == NULL) return NULL;
+  if (!element->head)
+    return 0;
+  else
+  {
+    element->head = false;
+    switch_state(element, element->allocated);
+    return 1;
+  }
 }
 
-struct allocator *global_alloc() {
-    global_allocator.init = global_init;
-    global_allocator.malloc = global_malloc;
-    global_allocator.realloc = global_realloc;
-    global_allocator.free = global_free;
-    return &global_allocator;
+/** Allocates a chunk of memory of the given size in global memory */
+static void *global_malloc(size_t size)
+{
+  struct Node *head = find_space(size);
+  if(head == NULL) return NULL;
+  head->allocated = size;
+  head->head = true;
+  switch_state(head, size);
+
+  return head->pool_element;
+}
+
+/** Resizes a chunk of memory in global memory previously allocated by malloc */
+static void *global_realloc(void *p, size_t size)
```

Comment:
`global_realloc` seems to have strange, undesired behavior. See the following example program for details.
```c
int main() {
    global_init();
    int *arr = global_malloc(sizeof(int)*3);
    arr[0] = 0;
    arr[1] = 1;
    arr[2] = 2;
    arr = global_realloc(arr, sizeof(int)*5);
    arr[3] = 3;
    arr[4] = 4;
    printf("%d\n", arr[0]);
    printf("%d\n", arr[1]);
    printf("%d\n", arr[2]);
    printf("%d\n", arr[3]);
    printf("%d\n", arr[4]);
    global_free(arr);
}
```
```
6717440
1
12
3
4
```